Introduce derived and extended Profiles for containers

yuumasato opened the pull request:
This is a protoype of build system to ~generate a separate datastream~ _derive a Profile_ for container variant of products.

~Instead of using CPEs to separate Rules that apply only to containers or machines, define a new product called rhel7-container. This will mean a separate source datastream for container and we will be able to just mark Rules applicable for containers with `prodtype="rhel7-container"`.~

This will avoid the problem of binding a Rule that can be applied for both machines and containers to only one of CPE.
A scenario easy to grasp is Rule `service_sshd_enabled`.
A Profile with such rule is okay to apply on VM or bare metal. Containers on the other hand don't need ssh, and it is also not recommended, but they can run it if really wanted.
We would like to, by default, not apply `service_sshd_enabled` to containers, but still allow users to tailor the Profile and add it.
By adding CPEs to Rules it is not possible. CPEs cannot be tailored.

Wtih a separate ~datastream and~ Profiles for containers user can tailor it and add rules to be applied to containers.

pep8speaks commented:
Hello @yuumasato! Thanks for submitting the PR.

 - In the file [`shared/utils/cpe-generate.py`](https://github.com/OpenSCAP/scap-security-guide/blob/a2ad3a7856a413bab5d1bc4554e0bc768ca68ce4/shared/utils/cpe-generate.py), following are the PEP8 issues :

> [Line 16:1](https://github.com/OpenSCAP/scap-security-guide/blob/a2ad3a7856a413bab5d1bc4554e0bc768ca68ce4/shared/utils/cpe-generate.py#L16): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 78:14](https://github.com/OpenSCAP/scap-security-guide/blob/a2ad3a7856a413bab5d1bc4554e0bc768ca68ce4/shared/utils/cpe-generate.py#L78): [E211](https://duckduckgo.com/?q=pep8%20E211) whitespace before '('
> [Line 168:80](https://github.com/OpenSCAP/scap-security-guide/blob/a2ad3a7856a413bab5d1bc4554e0bc768ca68ce4/shared/utils/cpe-generate.py#L168): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 170:80](https://github.com/OpenSCAP/scap-security-guide/blob/a2ad3a7856a413bab5d1bc4554e0bc768ca68ce4/shared/utils/cpe-generate.py#L170): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 180:80](https://github.com/OpenSCAP/scap-security-guide/blob/a2ad3a7856a413bab5d1bc4554e0bc768ca68ce4/shared/utils/cpe-generate.py#L180): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 182:80](https://github.com/OpenSCAP/scap-security-guide/blob/a2ad3a7856a413bab5d1bc4554e0bc768ca68ce4/shared/utils/cpe-generate.py#L182): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 184:80](https://github.com/OpenSCAP/scap-security-guide/blob/a2ad3a7856a413bab5d1bc4554e0bc768ca68ce4/shared/utils/cpe-generate.py#L184): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 187:80](https://github.com/OpenSCAP/scap-security-guide/blob/a2ad3a7856a413bab5d1bc4554e0bc768ca68ce4/shared/utils/cpe-generate.py#L187): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 197:80](https://github.com/OpenSCAP/scap-security-guide/blob/a2ad3a7856a413bab5d1bc4554e0bc768ca68ce4/shared/utils/cpe-generate.py#L197): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 216:80](https://github.com/OpenSCAP/scap-security-guide/blob/a2ad3a7856a413bab5d1bc4554e0bc768ca68ce4/shared/utils/cpe-generate.py#L216): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 218:80](https://github.com/OpenSCAP/scap-security-guide/blob/a2ad3a7856a413bab5d1bc4554e0bc768ca68ce4/shared/utils/cpe-generate.py#L218): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 222:1](https://github.com/OpenSCAP/scap-security-guide/blob/a2ad3a7856a413bab5d1bc4554e0bc768ca68ce4/shared/utils/cpe-generate.py#L222): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1

redhatrises commented:
I maybe missing something here, but why not just have a separate profile which seems similar? Why a separate product as we could have n-container products?

shawndwells commented:
On 8/10/17 5:37 PM, redhatrises wrote:
> I maybe missing something here, but why not just have a separate
> profile which seems similar? Why a separate product as we could have
> n-container products?

+1. Shipping multiple datastreams would be chaos.

yuumasato commented:
> +1. Shipping multiple datastreams would be chaos.

I see. And I have realized that with separate datastream it will not possible to tailor Profile and add Rules not applied to containers by default.

Another approach is for the build system to generate two profiles, one as usual, and one for containers. The build system would then separate what is applicable for containers.

redhatrises commented:
> Another approach is for the build system to generate two profiles, one as usual, and one for containers. The build system would then separate what is applicable for containers.

Still think you are complicating this a little. Create a profile under `RHEL/7/input/profiles/` called `stig-containers-upstream.xml` or something similar which extends="stig-rhel7-disa". In the new profile, enable or disable checks that do or do not apply to containers e.g. `<select idref="service_sshd_enabled" selected="false" />`, `<select idref="service_sshd_disabled" selected="true" />`, etc.

shawndwells commented:
On 8/11/17 9:57 AM, redhatrises wrote:
>
>     Another approach is for the build system to generate two profiles,
>     one as usual, and one for containers. The build system would then
>     separate what is applicable for containers.
>
> Still think you are complicating this a little. Create a profile under
> |RHEL/7/input/profiles/| called |stig-containers-upstream.xml| or
> something similar which extends="stig-rhel7-disa". In the new profile,
> enable or disable checks that do or do not apply to containers e.g.
> |<select idref="service_sshd_enabled" selected="false" />|, |<select
> idref="service_sshd_disabled" selected="true" />|, etc.
>

I'm still missing the need to do even that.  The OS SRG requirements are
going to be the same whether RHEL is bare metal, container, VM....

Only checks for installed services get evaluated. So things around SSH
will automatically show up as notapplicable.

Other elements, e.g. sysctls selections, may differ inside containers.
Generally those are templated though. Why not update the OVAL to pass if
inside a container?

redhatrises commented:
>>On 8/11/17 9:57 AM, redhatrises wrote:
>>
>> Another approach is for the build system to generate two profiles,
>> one as usual, and one for containers. The build system would then
>> separate what is applicable for containers.
>>
>> Still think you are complicating this a little. Create a profile under
>> `RHEL/7/input/profiles/` called `stig-containers-upstream.xml` or
>> something similar which extends="stig-rhel7-disa". In the new profile,
>> enable or disable checks that do or do not apply to containers e.g.
>> `<select idref="service_sshd_enabled" selected="false" />`, `<select
>> idref="service_sshd_disabled" selected="true" />`, etc.
>>
>
> I'm still missing the need to do even that.  The OS SRG requirements are
going to be the same whether RHEL is bare metal, container, VM....
>
> Only checks for installed services get evaluated. So things around SSH
will automatically show up as notapplicable.
>
> Other elements, e.g. sysctls selections, may differ inside containers.
> Generally those are templated though. Why not update the OVAL to pass if
> inside a container?

It's not a bad idea, but what it you wanted to enable or disable some of the existing rules for a container that you are checking? What if you want to check that sshd is disabled which is considered a best security practice for containers?

yuumasato commented:
Implemented prototype of proposed approach, XD.

A container Profile is generated, deriving from original Profile and extending `container-base` Profile.
I imagine this "container-base" Profile selecting Rules that are checking basic things in containers, that is usually not included in normal Profiles. It is like a `standard` profile for containers.

> Still think you are complicating this a little. Create a profile under RHEL/7/input/profiles/ called stig-containers-upstream.xml or something similar which extends="stig-rhel7-disa". In the new profile, enable or disable checks that do or do not apply to containers e.g. <select idref="service_sshd_enabled" selected="false" />, <select idref="service_sshd_disabled" selected="true" />, etc.

I would like to automate that using information in individual Rule, avoid manually disabling/enabling Rules in each "containerized" Profile, also Rules included in machine/container Profile can get out of sync easily.

IMHO it is much easier to look at each Rule and see if it is good fit for containers than going through each Profile and removing Rules. 
We were already doing it, marking each Rule with a CPE for machines, we are just changing it a little bit.

`cpe:/a:machine` was a definite decision, and the Rule could not be applied to containers, with `<machine /> the Rule is not recommended for containers, but user can still apply it.

mpreisler commented on shared/xccdf/services/base.xml:
```diff
@@ -117,7 +117,7 @@ multiple processors.
 <rationale>In an environment with multiple processors (now common), the irqbalance service
 provides potential speedups for handling interrupt requests.</rationale>
-<platform idref="cpe:/a:machine"/>
+<machine />
```
I am not a big fan of this syntax. It's not self-documenting at all. It doesn't allow to set some rules as container only. It doesn't allow anything for VMs.

mpreisler replied:
How about something like `<platform-restriction/>` which defaults to machine=true, container=true, vm=true and you can override any of them?

`<platform-restriction container="false"/>`

It's not great and something even a little better is required but it communications that there is some restriction on where this rule is to be used. And `platform-restriction` is not the right name for this element.

yuumasato replied:
How about `environment-restriction`? I have made updates using this name...
I like the idea to use this element to filter for multiple environments.

mpreisler commented on shared/utils/filter-machine-rules.py:
```diff
@@ -0,0 +1,67 @@
+
+def main():
+    xccdf_file = sys.argv[1]
+    xccdf_out_file = sys.argv[2]
```
argparse please...

yuumasato replied:
Using argparse now.

mpreisler commented on shared/utils/filter-machine-rules.py:
```diff
@@ -0,0 +1,67 @@
+import sys
+
+def parse_xml_file(xmlfile):
+    with open(xmlfile, 'r') as xml_file:
```
what is this function for? there is a function in ElementTree that does this and does it way better. `ElementTree.parse`.

yuumasato replied:
Function removed.

mpreisler commented on shared/utils/filter-machine-rules.py:
```diff
@@ -0,0 +1,67 @@
+                        % (profile_id, select_idref))
+                continue
+
+            x = rule.find('./machine')
```
this is a useless variable, use the method call directly in the condition

yuumasato replied:
Fixed.

yuumasato commented:
- Rebased  to fix conflicts.
- Changed filtering element
- Extended filtering capability to filter out Rules not recommended for machines in normal Profiles.

dahaic commented:
Jenkins test this please.

yuumasato commented:
I don't consider this WIP anymore.

yuumasato commented:
I'm closing this...

We have discussed this within the team and we won't be going down this path.

The main drawback is that it adds clutter into list of Profiles and the confusion it may create.
We will improve upon current solution and keep selecting content for containers by using CPEs.
